erizo_controller/erizoClient/src/webrtc-stacks/BaseStack.js

```diff
@@ -148,130 +117,337 @@ const BaseStack = (specInput) => {
+        }
+      }),
+      addStream: negotiationQueue.protectFunction((stream) => {
+        negotiationQueue.startEnqueuing();
+        that.peerConnectionFsm.addStream(stream).catch(onFsmError.bind(this));
+      }),
+      removeStream: negotiationQueue.protectFunction((stream) => {
+        negotiationQueue.startEnqueuing();
+        that.peerConnectionFsm.removeStream(stream).catch(onFsmError.bind(this));
+      }),
+      close: negotiationQueue.protectFunction(() => {
+        negotiationQueue.startEnqueuing();
+        that.peerConnectionFsm.close().catch(onFsmError.bind(this));
+      }),
+    },
+
+    firstLocalDescriptionQueue: {
+      createOffer:
+      firstLocalDescriptionQueue.protectFunction((isSubscribe = false,
+        forceOfferToReceive = false) => {
+        if (!firstLocalDescriptionSet) {
+          firstLocalDescriptionQueue.startEnqueuing();
+        }
+        if (!isSubscribe && !forceOfferToReceive) {
+          that.mediaConstraints = {
+            offerToReceiveVideo: false,
+            offerToReceiveAudio: false,
+          };
+        }
+        that.enqueuedCalls.negotiationQueue.createOffer(isSubscribe);
+      }),
 
-  const processOffer = firstLocalDescriptionQueue.protectFunction((message) => {
-    if (!firstLocalDescriptionSet) {
-      firstLocalDescriptionQueue.startEnqueuing();
-    }
-    _processOffer(message);
-  });
+      sendOffer: firstLocalDescriptionQueue.protectFunction(() => {
+        if (!firstLocalDescriptionSet) {
+          that.createOffer(true, true);
+          return;
+        }
+        setLocalDescForOffer(true, localDesc);
+      }),
 
-  const processAnswer = negotiationQueue.protectFunction((message) => {
-    const msg = message;
+      processOffer:
+      firstLocalDescriptionQueue.protectFunction((message) => {
+        if (!firstLocalDescriptionSet) {
+          firstLocalDescriptionQueue.startEnqueuing();
+        }
+        that.enqueuedCalls.negotiationQueue.processOffer(message);
+      }),
+    },
+  };
 
-    remoteSdp = SemanticSdp.SDPInfo.processString(msg.sdp);
-    const sessionVersion = remoteSdp && remoteSdp.origin && remoteSdp.origin.sessionVersion;
-    if (latestSessionVersion >= sessionVersion) {
-      Logger.warning(`processAnswer discarding old sdp, sessionVersion: ${sessionVersion}, latestSessionVersion: ${latestSessionVersion}`);
-      return;
-    }
-    negotiationQueue.startEnqueuing();
-    latestSessionVersion = sessionVersion;
-    Logger.info('Set remote and local description');
+  // Functions that are protected by the FSM.
+  // The promise of one has to be resolved before another can be called.
+  that.protectedCalls = {
+    protectedAddStream: (stream) => {
+      that.peerConnection.addStream(stream);
+      setTimeout(() => {
+        negotiationQueue.stopEnqueuing();
+        negotiationQueue.nextInQueue();
+      }, 0);
+      return Promise.resolve();
+    },
+    protectedRemoveStream: (stream) => {
```

Comment:
I'd keep a blank line between functions in the protectedCalls and queuedCalls, to improve readability a bit